sdk/monitor/Azure.Monitory.Query/src/TypeBinder.cs

```diff
@@ -0,0 +1,146 @@
+                _cache[type] = info = new BoundTypeInfo(type);
+            }
+
+            return info;
+        }
+
+
+        public abstract void Set<T>(TExchange destination, T value, BoundMemberInfo memberInfo);
+        public abstract bool TryGet<T>(BoundMemberInfo memberInfo, TExchange source,  out T value);
+    }
+    internal class BoundTypeInfo
+    {
+        public BoundTypeInfo(Type type)
+        {
+            List<BoundMemberInfo> members = new List<BoundMemberInfo>();
+            foreach (var memberInfo in type.GetMembers(BindingFlags.Public | BindingFlags.Instance))
+            {
+                switch (memberInfo)
+                {
+                    case PropertyInfo propertyInfo:
+                        members.Add((BoundMemberInfo) Activator.CreateInstance(typeof(BoundMemberInfo<>).MakeGenericType(propertyInfo.PropertyType), propertyInfo));
+                        break;
+                    case FieldInfo fieldInfo:
+                        members.Add((BoundMemberInfo) Activator.CreateInstance(typeof(BoundMemberInfo<>).MakeGenericType(fieldInfo.FieldType), fieldInfo));
+                        break;
+                }
+            }
+
+            Members = members.ToArray();
+        }
+
+        public BoundMemberInfo[] Members { get; }
+
+        public void Serialize<TExchange>(object o, TExchange destination, TypeBinder<TExchange> binderImplementation)
+        {
+            foreach (var member in Members)
+            {
+                if (member.CanRead)
+                {
+                    member.Serialize(o, destination, binderImplementation);
+                }
+            }
+        }
+
+        public void Deserialize<TExchange>(TExchange source, object o, TypeBinder<TExchange> binderImplementation)
+        {
+            foreach (var member in Members)
+            {
+                if (member.CanWrite)
+                {
+                    member.Deserialize(source, o, binderImplementation);
+                }
+            }
+        }
+    }
+
+    internal abstract class BoundMemberInfo
+    {
+        public BoundMemberInfo(MemberInfo memberInfo)
+        {
+            MemberInfo = memberInfo;
+        }
+
+        public string Name => MemberInfo.Name;
+        public MemberInfo MemberInfo { get; }
+        public abstract bool CanRead { get; }
+        public abstract bool CanWrite { get; }
+        public abstract void Serialize<TExchange>(object o, TExchange destination, TypeBinder<TExchange> binderImplementation);
+        public abstract void Deserialize<TExchange>(TExchange source, object o, TypeBinder<TExchange> binderImplementation);
+    }
+
+    public delegate T PropertyGetter<T>(object o);
+    public delegate void PropertySetter<T>(object o, T value);
+
+    internal sealed class BoundMemberInfo<T>: BoundMemberInfo
+    {
+        public BoundMemberInfo(PropertyInfo propertyInfo) : base(propertyInfo)
+        {
+            Getter = o => (T) propertyInfo.GetValue(o);
```

Comment:
We'll replace this part with IL Emit

Reply:
I went with Expressions for now. While direct IL emit would be more efficient but it requires to cross compile library assemblies that I would like to avoid for now.